returnn/datasets/basic.py

```diff
@@ -114,6 +114,7 @@ def __init__(self, name=None,
     self.rnd_seq_drop = None  # type: typing.Optional[Random]
     self.num_inputs = 0  # usually not used, but num_outputs instead, which is more generic
     self.num_outputs = None  # type: typing.Optional[typing.Dict[str,typing.Tuple[int,int]]]  # tuple is num-classes, len(shape).  # nopep8
+    self.default_data_key = "data"  # type: str  # Allows Datasets to re-define their input/default data key
```

Comment:
Why does the base `Dataset` class need this?
(And don't just explain here in the comment. This should be explained in the code. What is this used for?)

Reply:
From later usage in the code, it seems like this is used as the default seq length data key, e.g. for sorting. Then better call it `seq_length_data_key` or so. Why anway have "default" in the name?

Reply:
Also, why is this not configurable? (I.e. not a param in `__init__` here in the base class?)

Reply:
I call it default, because the sequence sorting is not the only potential usage. I did not make it configurable, because then I would need to make sure that all datasets treat this correctly when changed by the user. As "data" as default data key is used in more than 100 locations across all datasets, this would be a lot of work. If you want, I can extend the docstring (although for me the explanation is quite obvious), but I would not make this configurable for all datasets, but only for datasets which actually treat this correctly (in this case, the MetaDataset).

Reply:
> I call it default, because the sequence sorting is not the only potential usage.

I don't understand. What are potential usages? We don't really need this I think. And we should be more explicit. If there ever comes up another usage, I think it would be more clear to just have a separate flag for it. But I think it would really be bad if there would be other use cases where one data stream is somewhat special.

> I did not make it configurable, because then I would need to make sure that all datasets treat this correctly when changed by the user.

I don't really understand. If this is not configurable, it cannot really be used at all. Now it looks like you can just configure it for the MetaDataset. But actually it would be relevant to configure for the childs which you use in MetaDataset, e.g. HDFDataset, ExternSprintDataset, whatever.

> As "data" as default data key is used in more than 100 locations across all datasets, this would be a lot of work.

I doubt that there are so much hardcoded usages of "data". And it shouldn't really be much work. Basically (almost) just a global replace `"data"` by `seq_length_sort_data_key`.

> If you want, I can extend the docstring (although for me the explanation is quite obvious),

For me it is not obvious at all. If you call it `seq_length_sort_data_key`, and then shortly explain it is used for sorting the sequences, then it would be clear.

> but I would not make this configurable for all datasets, but only for datasets which actually treat this correctly (in this case, the MetaDataset).

But this doesn't really make sense, as explained.

Reply:
> But I think it would really be bad if there would be other use cases where one data stream is somewhat special.

Okay, but fixing this is definitely beyond the scope of this PR. 

> But actually it would be relevant to configure for the childs

This would require to make it configurable for all childs, which again is beyond the scope of this PR. I agree that I chose the "easy" way (for now) by setting this fixed in the base class. But to e.g. support `seq_length_sort_data_key` for the HDFDataset would require the changes currently on review in #438. I would have to check (and maybe change) all datasets.